internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -0,0 +1,585 @@
+	// was passed to this Compose method. If this call to Apply changes the XR
+	// in the API server (i.e. if it's not a no-op) the xr object that was
+	// passed to this method will have a stale meta.resourceVersion. This
+	// Subsequent attempts to update that object will therefore fail. This
+	// should be okay; the caller should keep trying until this is a no-op.
+	if err := c.client.Apply(ctx, xr); err != nil {
+		// It's important we don't proceed if this fails, because we need to be
+		// sure we've persisted our resource references before we create any new
+		// composed resources below.
+		return CompositionResult{}, errors.Wrap(err, errApplyXR)
+	}
+
+	// We apply all of our desired resources before we observe them in the loop
+	// below. This ensures that issues observing and processing one composed
+	// resource won't block the application of another.
+	for name, cd := range desired {
+		// TODO(negz): There's currently no way to control array/object merge
+		// behavior for resources from the Function pipeline. Hopefully this
+		// goes away entirely, and is replaced with server-side-apply.
+		// https://github.com/crossplane/crossplane/issues/4047
+		if err := c.client.Apply(ctx, cd.Resource, resource.MustBeControllableBy(xr.GetUID())); err != nil {
+			return CompositionResult{}, errors.Wrapf(err, errFmtApplyCD, name)
+		}
+	}
+
+	// Produce our array of resources resources to return to the Reconciler. The
+	// Reconciler uses this array to determine whether the XR is ready.
+	resources := make([]ComposedResource, 0, len(desired))
+	for name, r := range desired {
+		resources = append(resources, ComposedResource{ResourceName: name, Ready: r.Ready})
+	}
+
+	return CompositionResult{ConnectionDetails: xrConnDetails, Composed: resources, Events: events}, nil
+}
+
+// An ExistingComposedResourceObserver uses an XR's resource references to load
+// any existing composed resources from the API server. It also loads their
+// connection details.
+type ExistingComposedResourceObserver struct {
+	resource client.Reader
+	details  managed.ConnectionDetailsFetcher
+}
+
+// NewExistingComposedResourceObserver returns a ComposedResourceGetter that
+// fetches an XR's existing composed resources.
+func NewExistingComposedResourceObserver(c client.Reader, f managed.ConnectionDetailsFetcher) *ExistingComposedResourceObserver {
+	return &ExistingComposedResourceObserver{resource: c, details: f}
+}
+
+// ObserveComposedResources begins building composed resource state by
+// fetching any existing composed resources referenced by the supplied composite
+// resource, as well as their connection details.
+func (g *ExistingComposedResourceObserver) ObserveComposedResources(ctx context.Context, xr resource.Composite) (ComposedResourceStates, error) {
+	ors := ComposedResourceStates{}
+
+	for _, ref := range xr.GetResourceReferences() {
+		// The PTComposer writes references to resources that it didn't actually
```

Comment:
Why do we care `PTComposer` here? I assume we won't support switching modes either.

Reply:
Per https://github.com/crossplane/crossplane/pull/4500#discussion_r1322116042 I do think we should support switching modes. Or at least, somehow migrating existing P&T based XRs over to Functions.

So this code is covering the case where:

* An XR was using the `PTComposer`.
* The XR had failed to render a resource.
* The XR was migrated to the `FunctionComposer` before it had a chance to render the resource.

This does seem like a pretty slim edge case. On the other hand it's not too hard to have this little guard here.

Reply:
I've updated the comment to mention this explicitly.

Reply:
This does make me realize that there's another little edge case that we now don't cover:

* An XR is using a `mode: Resources` Composition with anonymous templates.
* The XR migrates to a `mode: Pipeline` Composition (which requires named resources)

We'll cut over to using the `FunctionComposer`, which assumes all existing composed resources will have the `crossplane.io/composition-resource-name` annotation. However, the composed resources won't have that annotation and we'll return an error here.

I think the `FallbackForAnonymoustemplates` composer used to guard against this case. I could be convinced that this is enough of an edge case we should cover it with just documentation. i.e. Tell folks that if they want to migrate to Functions they must first make sure all their templates are named (the PTComposer already has code that handles "migrating" from anonymous templates to named templates).